File: bindings/swift/Engine.swift
// Thin FFI wrapper over libjd; pure Foundation + Libjd, no platform UI deps.
// Shared source, compiled into both the macOS (macos/) and iOS (ios/)
// frontends via their project.yml source paths — see bindings/swift/README.md.

import Foundation
import Libjd

final class Engine {
    private let ctx: OpaquePointer
    /// `jd_state_ptr` is fixed for the context's lifetime, so it's read once.
    private let statePtr: UnsafePointer<jd_state>

    init() {
        Engine.checkABI()
        guard let raw = jd_init() else {
            fatalError("jd_init failed (allocation failure)")
        }
        guard let sp = jd_state_ptr(raw) else {
            fatalError("jd_state_ptr returned NULL")
        }
        ctx = raw
        statePtr = sp
    }

    deinit {
        jd_deinit(ctx)
    }

    // MARK: - Mutating operations

    /// Feed one keystroke and return the resulting state.
    @discardableResult
    func pressKey(_ byte: UInt8) -> EngineState {
        jd_press_key(ctx, Int8(bitPattern: byte))
        return state
    }

    /// Undo the most recent trie descent, or close a punctuation window.
    /// Never produces a commit.
    @discardableResult
    func backspace() -> EngineState {
        jd_backspace(ctx)
        return state
    }

    /// Drop the in-flight composition and any recorded commit.
    @discardableResult
    func reset() -> EngineState {
        jd_reset(ctx)
        return state
    }

    /// Point the anchor at candidate `index`, so the engine's automatic commits
    /// follow what the user is looking at. Out-of-range indices are ignored.
    @discardableResult
    func setAnchor(_ index: UInt32) -> EngineState {
        jd_set_anchor(ctx, index)
        return state
    }

    // MARK: - Reads

    /// The engine's current state, without touching it.
    var state: EngineState { EngineState(statePtr.pointee) }

    /// Read the candidates at `[start, start + count)`. Returns fewer than
    /// `count` at the end of the list, and none when nothing is in flight.
    ///
    /// A pure read: it never moves the anchor and never invalidates anything
    /// returned earlier, so a candidate strip can prefetch as far ahead as it
    /// likes with no bookkeeping beyond remembering how far it got.
    func readRange(from start: UInt32, count: Int) -> [Candidate] {
        guard count > 0 else { return [] }
        var raw = [query_option](repeating: query_option(), count: count)
        let written = raw.withUnsafeMutableBufferPointer { buf in
            Int(jd_read_range(ctx, start, UInt32(count), buf.baseAddress, UInt32(count)))
        }
        return raw.prefix(written).map(Candidate.init(raw:))
    }

    // MARK: - ABI guard

    /// `jd.h` declares these structs by hand, so it can drift from the
    /// compiled library with no diagnostic — and a size mismatch would corrupt
    /// memory. One comparison per process; `static let` makes it lazy and
    /// thread-safe.
    ///
    /// Compared against `stride`, not `size`: Swift's `size` excludes tail
    /// padding, while the library reports C's `sizeof`, which includes it.
    /// `jd_state` has 4 such bytes on a 64-bit target (28 used, 32 total).
    private static let abiChecked: Void = {
        precondition(
            Int(jd_abi_layout(UInt32(JD_ABI_SIZEOF_STATE))) == MemoryLayout<jd_state>.stride,
            "libjd jd_state layout mismatch — core/include/jd.h is out of sync with the library"
        )
        precondition(
            Int(jd_abi_layout(UInt32(JD_ABI_SIZEOF_OPTION))) == MemoryLayout<query_option>.stride,
            "libjd query_option layout mismatch — core/include/jd.h is out of sync with the library"
        )
        precondition(
            Int(jd_abi_layout(UInt32(JD_ABI_HINT_CAP))) == Int(JD_HINT_CAP),
            "libjd hint capacity mismatch — core/include/jd.h is out of sync with the library"
        )
    }()

    private static func checkABI() {
        _ = abiChecked
    }
}

File: android/app/src/main/java/com/hronro/imejd/engine/Engine.kt
// Port of bindings/swift/Engine.swift — a thin FFI wrapper over libjd.
//
// Candidates are addressed by flat index, not by page. [readRange] is a pure
// read: it never changes what the engine's automatic commits resolve to, so a
// candidate strip can prefetch as far ahead as it likes. Tell the engine what
// the user actually sees with [setAnchor].
//
// Thread-safety mirrors libjd's contract: a single context must not be used
// from multiple threads concurrently. All calls here come from the IME's main
// thread, so no locking is needed.
package com.hronro.imejd.engine

import java.io.Closeable

class Engine : Closeable {
    // jd_init returns NULL on allocation failure; fail fast with a clear
    // message instead of passing a 0 handle into the native pointer casts.
    private var ctx: Long = run {
        val handle = nativeInit()
        check(handle != 0L) { "jd_init failed (allocation failure)" }
        handle
    }

    /** How many candidates one native call can return; [readRange] loops. */
    private val readChunk: Int = nativeReadChunk()

    // close() zeroes ctx; passing that through JNI would hand libjd a NULL
    // context and segfault the IME process. Fail with a clear error instead,
    // matching the JS binding's disposed poisoning.
    private fun requireCtx(): Long = ctx.also { check(it != 0L) { "Engine is closed" } }

    /** Feed one keystroke and return the resulting state. */
    fun pressKey(byte: Byte): EngineState = nativePressKey(requireCtx(), byte)

    /**
     * Undo the most recent trie descent, or close a punctuation window.
     * Never produces a commit.
     */
    fun backspace(): EngineState = nativeBackspace(requireCtx())

    /** Drop the in-flight composition and any recorded commit. */
    fun reset(): EngineState = nativeReset(requireCtx())

    /**
     * Point the anchor at candidate [index], so the engine's automatic commits
     * follow what the user is looking at. Out-of-range indices are ignored.
     */
    fun setAnchor(index: Int): EngineState = nativeSetAnchor(requireCtx(), index)

    /** The engine's current state, without touching it. */
    val state: EngineState get() = nativeState(requireCtx())

    /**
     * Read the candidates at `[start, start + count)`. Returns fewer than
     * [count] at the end of the list, and none when nothing is in flight.
     *
     * A pure read — it never moves the anchor, so prefetch freely.
     */
    fun readRange(start: Int, count: Int): List<Candidate> {
        if (count <= 0) return emptyList()
        val handle = requireCtx()
        val out = ArrayList<Candidate>(count)
        var at = start
        while (out.size < count) {
            val want = minOf(count - out.size, readChunk)
            val chunk = nativeReadRange(handle, at, want)
            if (chunk.isEmpty()) break
            out.addAll(chunk)
            at += chunk.size
            if (chunk.size < want) break // hit the end of the list
        }
        return out
    }

    override fun close() {
        if (ctx != 0L) {
            nativeDeinit(ctx)
            ctx = 0L
        }
    }

    private external fun nativeInit(): Long
    private external fun nativeDeinit(ctx: Long)
    private external fun nativePressKey(ctx: Long, key: Byte): EngineState
    private external fun nativeBackspace(ctx: Long): EngineState
    private external fun nativeReset(ctx: Long): EngineState
    private external fun nativeSetAnchor(ctx: Long, index: Int): EngineState
    private external fun nativeState(ctx: Long): EngineState
    private external fun nativeReadRange(ctx: Long, start: Int, count: Int): List<Candidate>
    private external fun nativeReadChunk(): Int

    companion object {
        init {
            // Load ONLY the shim. Its DT_NEEDED pulls in libjd.so together with
            // libc.so as one load group, so libjd.so's libc references (e.g.
            // getauxval) resolve. Loading libjd.so on its own would fail — being
            // libc-free, it declares no NEEDED libc.so to resolve those against.
            //
            // JNI_OnLoad also verifies libjd's ABI layout against jd.h and
            // fails the load on a mismatch.
            System.loadLibrary("jdjni")
        }
    }
}
